quaterion/train/xbm/xbm_buffer.py

```diff
@@ -0,0 +1,62 @@
+from typing import Tuple
+
+import torch
+from torch import LongTensor, Tensor
+
+from .xbm_config import XbmConfig, XbmDevice
+
+
+class XbmBuffer:
+    """A buffer implementation to hold recent N embeddings and target values.
+
+    Inspired by https://github.com/msight-tech/research-xbm/blob/master/ret_benchmark/modeling/xbm.py
+
+    Args:
+        config: Config class to configure XBM settings.
+    """
+
+    def __init__(self, config: XbmConfig):
+        if config.device == XbmDevice.AUTO:
```

Comment:
I'm afraid it's `CacheType` whose behavior should be changed because [it's not reliable to get device from a module](https://stackoverflow.com/questions/58926054/how-to-get-the-device-type-of-a-pytorch-module-conveniently).

Others are good catches --already fixes some of them, and I'll continue to fix others tomorrow.